service/src/main/java/bio/terra/workspace/service/iam/ControlledResourceSamPolicyBuilder.java

```diff
@@ -0,0 +1,206 @@
+            policyMap.get(ControlledResourceIamRole.EDITOR), userRequest);
+        // if we have an assigned user, set up their permission
+        if (privateUser != null) {
+          policyMap.get(privateIamRole).addMemberEmailsItem(privateUser);
+        }
+        applyPolicyMap(request, policyMap);
+        break;
+    }
+  }
+
+  /**
+   * For the private resources we always fill in READER, WRITER, and EDITOR even if empty. This
+   * generates a map by role of the policy.
+   */
+  private Map<ControlledResourceIamRole, AccessPolicyMembershipV2> makePolicyMap() {
+    Map<ControlledResourceIamRole, AccessPolicyMembershipV2> policyMap = new HashMap<>();
+    policyMap.put(
+        ControlledResourceIamRole.READER,
+        new AccessPolicyMembershipV2().addRolesItem(ControlledResourceIamRole.READER.toSamRole()));
+    policyMap.put(
+        ControlledResourceIamRole.WRITER,
+        new AccessPolicyMembershipV2().addRolesItem(ControlledResourceIamRole.WRITER.toSamRole()));
+    policyMap.put(
+        ControlledResourceIamRole.EDITOR,
+        new AccessPolicyMembershipV2().addRolesItem(ControlledResourceIamRole.EDITOR.toSamRole()));
+    return policyMap;
+  }
+
+  /**
+   * Given the map created in @{@link #makePolicyMap()}, populate the resource request with the
+   * policies.
+   *
+   * @param request resource request for setting the roles on the Sam resource creation
+   * @param policyMap previously created policy map
+   */
+  private void applyPolicyMap(
+      CreateResourceRequestV2 request,
+      Map<ControlledResourceIamRole, AccessPolicyMembershipV2> policyMap) {
+    for (Entry<ControlledResourceIamRole, AccessPolicyMembershipV2> entry : policyMap.entrySet()) {
+      request.putPoliciesItem(entry.getKey().toSamRole(), entry.getValue());
+    }
+  }
+
+  /**
+   * Add WSM's service account as the owner of a controlled resource in Sam. Used for admin
+   * reassignment of resources. This assumes samService.initialize() has already been called, which
+   * should happen on start.
+   */
+  private void addWsmResourceOwnerPolicy(CreateResourceRequestV2 request)
+      throws InterruptedException {
+    try {
+      AuthenticatedUserRequest wsmRequest =
+          new AuthenticatedUserRequest().token(Optional.of(samService.getWsmServiceAccountToken()));
+      String wsmSaEmail = samService.getUserEmailFromSam(wsmRequest);
+      AccessPolicyMembershipV2 ownerPolicy =
+          new AccessPolicyMembershipV2()
+              .addRolesItem(ControlledResourceIamRole.OWNER.toSamRole())
+              .addMemberEmailsItem(wsmSaEmail);
+      request.putPoliciesItem(ControlledResourceIamRole.OWNER.toSamRole(), ownerPolicy);
+    } catch (InternalServerErrorException e) {
+      // In cases where WSM is not running as a service account (e.g. unit tests), the above call to
+      // get application default credentials will fail. This is fine, as those cases don't create
+      // real resources.
+      logger.warn(
```

Comment:
Was just thinking if this fails for other reasons in the real world, then we'd swallow this exception. Not sure what the risk/impact is though, so not a strong objection.

Reply:
True. It would be very bad if this happened in production. I'll make a ticket to track.